### ISSUE TITLE
gitlab-runner docker executor broken since golang 1.20.6

### ISSUE THREAD

**Issue by yayayayaka** [NixOS/nixpkgs#245365]
### Describe the bug
A new release in golang 1.20.6 (https://github.com/golang/go/issues/60374, https://github.com/NixOS/nixpkgs/pull/242949) fixed the validation of Host headers ([CVE-2023-29406](https://nvd.nist.gov/vuln/detail/CVE-2023-29406)). Unfortunately, the validation will now fail if the contents of the Host Header is a unix socket addess (as mentioned in https://github.com/golang/go/issues/61431 and https://github.com/moby/moby/issues/45935).

### Screenshots
Example CI run:
```
Running with gitlab-runner 16.1.0 (v16.1.0)
[...]
Preparing the "docker" executor
Using Docker executor with image alpine ...
Pulling docker image alpine ...
Using docker image [...]
Preparing environment
ERROR: Job failed (system failure): prepare environment: http: invalid Host header (exec.go:71:0s). Check https://docs.gitlab.com/runner/shells/index.html#shell-profile-loading for more information
```

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.4.3, NixOS, 23.05 (Stoat), 23.05pre-git`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.3`
 - nixpkgs: `/etc/src/nixpkgs`
```

**Comment by Janik-Haag** [comment]
Is this related to:  #244159 ?

**Comment by yayayayaka** [comment]
Yes, it is.

**Comment by pyrotek45** [comment]
this has ruined the ability to do programming work with dev containers on vscode. hope it gets fixed soon, anyone know a workaround for now?

**Comment by yayayayaka** [comment]
The only workaround I came up with so far was to rebuild gitlab-runner with Go 1.20.5 (which has security implications and would not recommend).

**Comment by svenstaro** [comment]
Reported upstream here: https://gitlab.com/gitlab-org/gitlab-runner/-/issues/36051

**Comment by angerman** [comment]
Is my understanding correct, that bumping nixpkgs past https://github.com/NixOS/nixpkgs/pull/245381, would get this back working?

**Comment by yayayayaka** [comment]
No, you would need to revert 3c53623bf432a069463b32d2ac806b0936b815f6

**Comment by angerman** [comment]
Looks like I failed to revert that via an overlay. Ohh, well full blown fork+patch then.

It's really annoying that we ended up in a situation where you can't have gitlab _and_ GitHub runners deployed from the same revision, as GitHub runners are _too_ old to work with current GitHub, and gitlab-runners build against go which is busted 🤣

**Comment by vidbina** [comment]
I've noticed the `http: invalid Host header` today when I was trying to use Docker to build some images and run some containers.

In https://github.com/moby/moby/pull/45942 it was mentioned that the issue pops up due to some changes in how host headers are handled since Go v1.19+.

My temporary workaround was to build docker with a golang version < v1.19 by setting `virtualization.docker.package` to use Docker which I overrode to build with Go v1.18.

```diff
modified   virt.nix
@@ -25,6 +25,9 @@
   virtualisation = {
     docker = {
       enable = true;
+      package = pkgs.bleeding.docker.override {
+        buildGoPackage = pkgs.buildGo118Package;
+      };
       daemon.settings = {
         features = {
           buildkit = true;
```

**Comment by luochen1990** [comment]
> Pass `host` to your docker runner in `config.toml`:
> 
> ```
> [[runners]]
> ...
> [runners.docker]
> ...
> host = "tcp://127.0.0.1:2375"
> ```

@codec Is this about this option: `services.gitlab-runner.services.<name>.dockerExtraHosts` ?  or how can I add nixos config to workaround?

**Comment by chvp** [comment]
@luochen1990 this is how I worked around it in my nixos config: https://github.com/chvp/nixos-config/commit/1c7eab46cc30babd7d8a3f7df7169146062f452e

**Comment by luochen1990** [comment]
@chvp Thanks very much!  but I got following error:

```
Preparing the "docker" executor
ERROR: Failed to remove network for build
ERROR: Preparation failed: proxy: unknown scheme: http
Will be retried in 3s ...
ERROR: Failed to remove network for build
ERROR: Preparation failed: proxy: unknown scheme: http
Will be retried in 3s ...
ERROR: Failed to remove network for build
ERROR: Preparation failed: proxy: unknown scheme: http
Will be retried in 3s ...
ERROR: Job failed (system failure): proxy: unknown scheme: http
```

seems it's the proxy settings cause this issue.

**Comment by felipelalli** [comment]
This thread must be closed as duplicated. #244159

**Comment by luochen1990** [comment]
Here is my workaround:

```
  #Workaround of https://github.com/NixOS/nixpkgs/issues/245365
  virtualisation.docker.listenOptions = [ "/run/docker.sock" "127.0.0.1:2375" ];
  systemd.services.gitlab-runner.environment.DOCKER_HOST = "tcp://127.0.0.1:2375";
  networking.proxy.allProxy = null; #Workaround of https://github.com/docker-archive/toolbox/issues/539
```

you might need `networking.proxy.allProxy = null;` or similar thing if you got "http proxy" issue.

**Comment by dminuoso** [comment]
> The only workaround I came up with so far was to rebuild gitlab-runner with Go 1.20.5 (which has security implications and would not recommend).

https://gitlab.com/gitlab-org/gitlab-runner/-/issues/36051#note_1494022403

1.20.5 is formally the best supported version. 


By the way,


```nix
{ ... }:

let
  dockerHost = "127.0.0.1:2375";
in {
  virtualisation.docker.listenOptions = [
    "/run/docker.sock"
    dockerHost
  ];
  systemd.services.gitlab-runner.environment.DOCKER_HOST = "tcp://${dockerHost}";
}
```

Seems to not work for me. Im still getting 

```
ERROR: Job failed (system failure): prepare environment: http: invalid Host header (exec.go:71:0s). Check https://docs.gitlab.com/runner/shells/index.html#shell-profile-loading for more information

```

**Comment by 5aaee9** [comment]
@dminuoso Try systemctl restart `gitlab-runner.service` and `docker.socket`

**Comment by unode** [comment]
Upstream merged https://gitlab.com/gitlab-org/gitlab-runner/-/merge_requests/4249 but hasn't landed on a release yet.

**Comment by takeda** [comment]
Would it be feasible to temporarily just apply this patch to our derivation?

https://gitlab.com/gitlab-org/gitlab-runner/-/merge_requests/4249.patch

**Comment by unode** [comment]
Agree with @takeda, specially considering that the next gitlab-runner release will happen in a little under a month (more if the patch isn't included for some reason).

**Comment by takeda** [comment]
I created PR including it: https://github.com/NixOS/nixpkgs/pull/250006 please note that I did not test if it resolves this issue.

**Comment by bachp** [comment]
Fixed by https://github.com/NixOS/nixpkgs/pull/249908

**Comment by CobaltCause** [comment]
That fixed it for me, though I had to reboot for it to take effect. Most likely I could've restarted some subset of things more granularly, but didn't really feel like figuring out what subset.

**Comment by chvp** [comment]
What's the status of this issue on 23.05?

**Comment by CobaltCause** [comment]
https://github.com/NixOS/nixpkgs/pull/249908 wasn't backported, so probably still an issue there.

**Comment by unode** [comment]
gitlab-runner 1.3.0 came out meanwhile. See #250521 for a follow up

**Comment by bachp** [comment]
16.3.0 is now available both in unstable and 23.05
